### DISCUSSION TITLE
Unexpected behavior when trying to (not) set .valueAsNumber

### DISCUSSION THREAD

**Discussion by Benexx** [lit/lit#4589] [Q&A] [answered]
### Which package(s) are affected?

Lit Core (lit / lit-html / lit-element / reactive-element)

### Description

II'm not sure it is a bug but I do have an unexpected behavior (in Firefox and Chromium) when making a webComponent that wraps a number input (in the full element we show some error messages). 
The goal is to no change the user input when he is changing the value (make the field empty, remove the focus or reset the cursor position). It works quite nice except when you preload the element with a `.value`. 

If you don't set the value and type `123-` by hand the input is not cleared. 
If the value is `123` and you add a `-` by hand the input is cleared. After this you could type `123-` without the input clearing. To reproduce the problem again you need to reload the element.

I tried to make the component as small as possible, we do use an internal variable to prevent unneeded renders for example but in some cases we do trigger a `requestUpdate` and that will clear the input.

My criteria:
- Allow setting the `.value` property on the custom element before the first render (default value)
- Allow `render()` to be triggered while the user is changing the value. We want to show some custom error texts while the user is changing the value.
- Allow changing the `.value` property after the user did change it

The element:
``` JS
import { html, LitElement } from 'lit';
import { customElement, property } from 'lit/decorators.js';
import { ifDefined } from 'lit/directives/if-defined.js';
import { live } from 'lit/directives/live.js';

@customElement('number-element')
export class NumberElementComponent extends LitElement {
    
    @property({ attribute: false })
    public accessor value: number;

    public render() {
        return html`
<input 
 type="number" 
 .valueAsNumber=${ifDefined(this.value != null ? live(this.value) : null)}
 @input=${this.handleInput} 
/>
`;
    }

    private handleInput(e: { currentTarget: HTMLInputElement, preventDefault() }): void {
        const value = e.currentTarget.valueAsNumber;
        this.value = isNaN(value) ? null : value;
    }
}
```

The usage:
```JS
html`<number-element .value=${123}></number-element>`
```

### Reproduction

The input should not be cleared when you replace or append a `-`: [Playground](https://lit.dev/playground/#project=W3sibmFtZSI6InJlcHJvLnRzIiwiY29udGVudCI6ImltcG9ydCB7IGh0bWwsIExpdEVsZW1lbnQgfSBmcm9tICdsaXQnO1xuaW1wb3J0IHsgY3VzdG9tRWxlbWVudCwgcHJvcGVydHkgfSBmcm9tICdsaXQvZGVjb3JhdG9ycy5qcyc7XG5pbXBvcnQgeyBpZkRlZmluZWQgfSBmcm9tICdsaXQvZGlyZWN0aXZlcy9pZi1kZWZpbmVkLmpzJztcbmltcG9ydCB7IGxpdmUgfSBmcm9tICdsaXQvZGlyZWN0aXZlcy9saXZlLmpzJztcblxuQGN1c3RvbUVsZW1lbnQoJ251bWJlci1lbGVtZW50JylcbmV4cG9ydCBjbGFzcyBOdW1iZXJFbGVtZW50Q29tcG9uZW50IGV4dGVuZHMgTGl0RWxlbWVudCB7XG4gICAgXG4gICAgQHByb3BlcnR5KHsgdHlwZTogJ251bWJlcid9KVxuICAgIHZhbHVlOiBudW1iZXI7XG5cbiAgICBwdWJsaWMgcmVuZGVyKCkge1xuICAgICAgICByZXR1cm4gaHRtbGBcbjxpbnB1dCBcbiB0eXBlPVwibnVtYmVyXCIgXG4gLnZhbHVlQXNOdW1iZXI9JHtpZkRlZmluZWQodGhpcy52YWx1ZSAhPSBudWxsID8gbGl2ZSh0aGlzLnZhbHVlKSA6IG51bGwpfVxuIEBpbnB1dD0ke3RoaXMuaGFuZGxlSW5wdXR9IFxuLz5cbmA7XG4gICAgfVxuXG4gICAgcHJpdmF0ZSBoYW5kbGVJbnB1dChlOiB7IGN1cnJlbnRUYXJnZXQ6IEhUTUxJbnB1dEVsZW1lbnQsIHByZXZlbnREZWZhdWx0KCkgfSk6IHZvaWQge1xuICAgICAgICBjb25zdCB2YWx1ZSA9IGUuY3VycmVudFRhcmdldC52YWx1ZUFzTnVtYmVyO1xuICAgICAgICB0aGlzLnZhbHVlID0gaXNOYU4odmFsdWUpID8gbnVsbCA6IHZhbHVlO1xuICAgIH1cbn1cblxuZGVjbGFyZSBnbG9iYWwge1xuICAgIGludGVyZmFjZSBIVE1MRWxlbWVudFRhZ05hbWVNYXAge1xuICAgICAgICAnbnVtYmVyLWVsZW1lbnQnOiBOdW1iZXJFbGVtZW50Q29tcG9uZW50O1xuICAgIH1cbn0ifSx7Im5hbWUiOiJpbmRleC5odG1sIiwiY29udGVudCI6IjwhRE9DVFlQRSBodG1sPlxuPGhlYWQ-XG4gIDxzY3JpcHQgdHlwZT1cIm1vZHVsZVwiIHNyYz1cIi4vcmVwcm8uanNcIj48L3NjcmlwdD5cbjwvaGVhZD5cbjxib2R5PlxuICA8bnVtYmVyLWVsZW1lbnQgdmFsdWU9MTIzPjwvbnVtYmVyLWVsZW1lbnQ-XG48L2JvZHk-XG4ifSx7Im5hbWUiOiJwYWNrYWdlLmpzb24iLCJjb250ZW50Ijoie1xuICBcImRlcGVuZGVuY2llc1wiOiB7XG4gICAgXCJsaXRcIjogXCJeMi4wLjBcIixcbiAgICBcIkBsaXQvcmVhY3RpdmUtZWxlbWVudFwiOiBcIl4xLjAuMFwiLFxuICAgIFwibGl0LWVsZW1lbnRcIjogXCJeMy4wLjBcIixcbiAgICBcImxpdC1odG1sXCI6IFwiXjIuMC4wXCJcbiAgfVxufSIsImhpZGRlbiI6dHJ1ZX1d)

### Workaround

We couldn't find any workarounds

### Is this a regression?

No or unsure. This never worked, or I haven't tried before.

### Affected versions

Fails for lit@3.1.0 and lit@3.1.2

### Browser/OS/Node environment

Firefox (123.0.1 (64-bit)) and Chrome (122.0.6261.129 (64-bit))

**Comment by augustjk** [comment]
Hmm.. is the idea to never modify what's being shown in the input from what the user is typing?

In that case you maybe don't even want to do a value property binding to the element. You can use the `value` _attribute_ to set the initial value. Then read it directly from the element when you need the current value that the user modified. No `input` event handling to keep it synced.

If you don't want to adjust the input while the user is typing, but _do_ want to modify it after the user clicks away, then you should handle the `change` event, rather than `input`, and you can keep the property binding.

If you do keep the property binding, `isDefined()` isn't really doing much for you there. It's main use is for attribute binding to actually remove the attribute, but for property binding it's equivalent to setting the value to `undefined` and you already have a ternary inside.

Because you're also coercing values resulting in `NaN` to `null`, Lit's change detection won't be able to tell when the user repeatedly puts in non-number values. The `live()` can also return a `noChange` sentinel value which I think is messing with the change detection in why you see the input not changing after manually typing the number and `-`.

**Comment by Benexx** [comment]
My criteria:
- Allow setting the `.value` property on the custom element before the first render (default value)
- Allow `render()` to be triggered while the user is changing the value. We want to show some custom error texts while the user is changing the value.
- Allow changing the `.value` property after the user did change it

I tried your suggestion ([playground](https://lit.dev/playground/#project=W3sibmFtZSI6InJlcHJvLnRzIiwiY29udGVudCI6ImltcG9ydCB7IGh0bWwsIExpdEVsZW1lbnQgfSBmcm9tICdsaXQnO1xuaW1wb3J0IHsgY3VzdG9tRWxlbWVudCwgcHJvcGVydHkgfSBmcm9tICdsaXQvZGVjb3JhdG9ycy5qcyc7XG5pbXBvcnQgeyBpZkRlZmluZWQgfSBmcm9tICdsaXQvZGlyZWN0aXZlcy9pZi1kZWZpbmVkLmpzJztcbmltcG9ydCB7IGxpdmUgfSBmcm9tICdsaXQvZGlyZWN0aXZlcy9saXZlLmpzJztcblxuZG9jdW1lbnQuZ2V0RWxlbWVudEJ5SWQoJ2J1dHRvbicpLmFkZEV2ZW50TGlzdGVuZXIoJ2NsaWNrJywgKCkgPT4ge1xuICBjb25zdCBuZXdWYWx1ZSA9IG5ldyBEYXRlKCkuZ2V0VGltZSgpIC8gMTA7XG4gIGNvbnNvbGUubG9nKCdMb2FkaW5nOicsIG5ld1ZhbHVlKTtcbiAgKGRvY3VtZW50LmdldEVsZW1lbnRCeUlkKCdlbGVtZW50JykgYXMgTnVtYmVyRWxlbWVudENvbXBvbmVudCkudmFsdWUgPSBuZXdWYWx1ZTtcbn0pO1xuXG5AY3VzdG9tRWxlbWVudCgnbnVtYmVyLWVsZW1lbnQnKVxuZXhwb3J0IGNsYXNzIE51bWJlckVsZW1lbnRDb21wb25lbnQgZXh0ZW5kcyBMaXRFbGVtZW50IHtcbiAgX3ZhbHVlOiBudW1iZXI7XG4gIF9sb2FkVmFsdWU6IHN0cmluZzsgIFxuICBcbiAgQHByb3BlcnR5KHsgdHlwZTogTnVtYmVyIH0pXG4gIHNldCB2YWx1ZSh2OiBudW1iZXIpIHtcbiAgICB0aGlzLl92YWx1ZSA9IHY7XG5cbiAgICBjb25zdCBvbGRMb2FkVmFsdWUgPSB0aGlzLl9sb2FkVmFsdWU7XG4gICAgdGhpcy5fbG9hZFZhbHVlID0gdiA9PSBudWxsIHx8IGlzTmFOKHYpID8gJycgOiB2LnRvU3RyaW5nKCkucmVwbGFjZSgnLCcsICcuJyk7XG5cbiAgICB0aGlzLnJlcXVlc3RVcGRhdGUoKTtcbiAgfVxuXG4gIGdldCB2YWx1ZSgpOiBudW1iZXIge1xuICAgIHJldHVybiB0aGlzLl92YWx1ZSA9PSBudWxsIHx8IGlzTmFOKHRoaXMuX3ZhbHVlKSA_IG51bGwgOiB0aGlzLl92YWx1ZTtcbiAgfVxuXG4gIHB1YmxpYyByZW5kZXIoKSB7XG4gICAgbGV0IGxvYWRWYWx1ZTogc3RyaW5nO1xuICAgIGlmKHRoaXMuX2xvYWRWYWx1ZSAhPSBudWxsKXtcbiAgICAgIGxvYWRWYWx1ZSA9IHRoaXMuX2xvYWRWYWx1ZTtcbiAgICAgIGNvbnNvbGUubG9nKCdSZW5kZXIgd2l0aCcsIGxvYWRWYWx1ZSk7XG4gICAgICBcbiAgICAgIHRoaXMuX2xvYWRWYWx1ZSA9IG51bGw7XG4gICAgfVxuXG4gICAgcmV0dXJuIGh0bWxgXG48aW5wdXQgXG4gdHlwZT1cIm51bWJlclwiIFxuIHZhbHVlPSR7aWZEZWZpbmVkKGxvYWRWYWx1ZSl9XG4gQGlucHV0PSR7dGhpcy5oYW5kbGVJbnB1dH0gXG4vPlxuYDtcbiAgfVxuXG4gIHByaXZhdGUgaGFuZGxlSW5wdXQoZTogeyBjdXJyZW50VGFyZ2V0OiBIVE1MSW5wdXRFbGVtZW50IH0pIHtcbiAgICB0aGlzLl92YWx1ZSA9IGUuY3VycmVudFRhcmdldC52YWx1ZUFzTnVtYmVyO1xuICB9XG59XG5cbmRlY2xhcmUgZ2xvYmFsIHtcbiAgaW50ZXJmYWNlIEhUTUxFbGVtZW50VGFnTmFtZU1hcCB7XG4gICAgJ251bWJlci1lbGVtZW50JzogTnVtYmVyRWxlbWVudENvbXBvbmVudDtcbiAgfVxufSJ9LHsibmFtZSI6ImluZGV4Lmh0bWwiLCJjb250ZW50IjoiPCFET0NUWVBFIGh0bWw-XG48aGVhZD5cbiAgPHNjcmlwdCB0eXBlPVwibW9kdWxlXCIgc3JjPVwiLi9yZXByby5qc1wiPjwvc2NyaXB0PlxuPC9oZWFkPlxuPGJvZHk-XG4gIDxudW1iZXItZWxlbWVudCBpZD1cImVsZW1lbnRcIiB2YWx1ZT1cIjEzMzdcIj48L251bWJlci1lbGVtZW50PlxuICA8YnV0dG9uIGlkPVwiYnV0dG9uXCI-VXBkYXRlIHZhbHVlPC9idXR0b24-XG48L2JvZHk-XG5cblxuIn0seyJuYW1lIjoicGFja2FnZS5qc29uIiwiY29udGVudCI6IntcbiAgXCJkZXBlbmRlbmNpZXNcIjoge1xuICAgIFwibGl0XCI6IFwiXjIuMC4wXCIsXG4gICAgXCJAbGl0L3JlYWN0aXZlLWVsZW1lbnRcIjogXCJeMS4wLjBcIixcbiAgICBcImxpdC1lbGVtZW50XCI6IFwiXjMuMC4wXCIsXG4gICAgXCJsaXQtaHRtbFwiOiBcIl4yLjAuMFwiXG4gIH1cbn0iLCJoaWRkZW4iOnRydWV9XQ)) but it seems that the `value` attribute is ignored after the first change by the user.

**Comment by Totati** [comment, marked as answer]
Haha, we had similar issues. [This is our solution.](https://lit.dev/playground/#project=W3sibmFtZSI6InJlcHJvLnRzIiwiY29udGVudCI6ImltcG9ydCB7IGh0bWwsIExpdEVsZW1lbnQsIG5vQ2hhbmdlIH0gZnJvbSAnbGl0JztcbmltcG9ydCB7IGN1c3RvbUVsZW1lbnQsIHByb3BlcnR5IH0gZnJvbSAnbGl0L2RlY29yYXRvcnMuanMnO1xuaW1wb3J0IHsgaWZEZWZpbmVkIH0gZnJvbSAnbGl0L2RpcmVjdGl2ZXMvaWYtZGVmaW5lZC5qcyc7XG5pbXBvcnQgeyBsaXZlIH0gZnJvbSAnbGl0L2RpcmVjdGl2ZXMvbGl2ZS5qcyc7XG5cbkBjdXN0b21FbGVtZW50KCdudW1iZXItZWxlbWVudCcpXG5leHBvcnQgY2xhc3MgTnVtYmVyRWxlbWVudENvbXBvbmVudCBleHRlbmRzIExpdEVsZW1lbnQge1xuICAgIFxuICAgIEBwcm9wZXJ0eSh7IHR5cGU6ICdudW1iZXInfSlcbiAgICB2YWx1ZTogbnVtYmVyO1xuXG4gICAgcHVibGljIHJlbmRlcigpIHtcbiAgICAgICAgcmV0dXJuIGh0bWxgXG48aW5wdXQgXG4gdHlwZT1cIm51bWJlclwiIFxuIC52YWx1ZUFzTnVtYmVyPSR7bGl2ZSh0aGlzLnZhbHVlID8_IG5vQ2hhbmdlKX1cbiBAaW5wdXQ9JHt0aGlzLmhhbmRsZUlucHV0fSBcbi8-JHt0aGlzLnZhbHVlfVxuYDtcbiAgICB9XG5cbiAgICBwcml2YXRlIGhhbmRsZUlucHV0KGU6IHsgY3VycmVudFRhcmdldDogSFRNTElucHV0RWxlbWVudCwgcHJldmVudERlZmF1bHQoKSB9KTogdm9pZCB7XG4gICAgICAgIGNvbnN0IHZhbHVlID0gZS5jdXJyZW50VGFyZ2V0LnZhbHVlQXNOdW1iZXI7XG4gICAgICAgIHRoaXMudmFsdWUgPSBpc05hTih2YWx1ZSkgPyBudWxsIDogdmFsdWU7XG4gICAgfVxufVxuXG5kZWNsYXJlIGdsb2JhbCB7XG4gICAgaW50ZXJmYWNlIEhUTUxFbGVtZW50VGFnTmFtZU1hcCB7XG4gICAgICAgICdudW1iZXItZWxlbWVudCc6IE51bWJlckVsZW1lbnRDb21wb25lbnQ7XG4gICAgfVxufSJ9LHsibmFtZSI6ImluZGV4Lmh0bWwiLCJjb250ZW50IjoiPCFET0NUWVBFIGh0bWw-XG48aGVhZD5cbiAgPHNjcmlwdCB0eXBlPVwibW9kdWxlXCIgc3JjPVwiLi9yZXByby5qc1wiPjwvc2NyaXB0PlxuPC9oZWFkPlxuPGJvZHk-XG4gIDxudW1iZXItZWxlbWVudCB2YWx1ZT0xMjM-PC9udW1iZXItZWxlbWVudD5cbjwvYm9keT5cbiJ9LHsibmFtZSI6InBhY2thZ2UuanNvbiIsImNvbnRlbnQiOiJ7XG4gIFwiZGVwZW5kZW5jaWVzXCI6IHtcbiAgICBcImxpdFwiOiBcIl4yLjAuMFwiLFxuICAgIFwiQGxpdC9yZWFjdGl2ZS1lbGVtZW50XCI6IFwiXjEuMC4wXCIsXG4gICAgXCJsaXQtZWxlbWVudFwiOiBcIl4zLjAuMFwiLFxuICAgIFwibGl0LWh0bWxcIjogXCJeMi4wLjBcIlxuICB9XG59IiwiaGlkZGVuIjp0cnVlfV0)
![image](https://github.com/lit/lit/assets/17689719/8e5a5d2e-9f94-4a6e-94a0-6d11096aaeee)
[Base on the source](https://github.com/lit/lit/blob/dd74c604b859e432159c23c3c99f36fde3fe147a/packages/lit-html/src/directives/live.ts#L41)

**Comment by justinfagnani** [comment]
A couple of notes:
- `.value` isn't typed correctly. You have it typed as number, but you set it to `null`, so it should at least be `number | null`.
- You're setting `input.valueAsNumber` to `null` when `this.value` is `null` and that definitely will clear the input. That much seems expected.
- `ifDefined()` doesn't really do anything for property bindings
- Directives should usually be the only expression in a binding. If a binding value switches to and from a directive result, the state of the directive will be lost. You almost always want `${live(...)}`
- The `type` property option should be a JS primitive constructor, not a string. `'number'` won't do any convertion of the attribute to a number. `Number` will.

I think one confusing think about this component is that right now the input is "controlled" (meaning it's always set from the outside) when the value is valid, and I think you're trying to make it "uncontrolled" (meaning it keeps it's own state) when the value is invalid.

My guess is that's more difficult to reason about than keeping the input always controlled, but @Totati's suggestion works if you keep it that way.  See [this updated demo](https://lit.dev/playground/#project=W3sibmFtZSI6InJlcHJvLnRzIiwiY29udGVudCI6ImltcG9ydCB7aHRtbCwgTGl0RWxlbWVudCwgbm9DaGFuZ2V9IGZyb20gJ2xpdCc7XG5pbXBvcnQge2N1c3RvbUVsZW1lbnQsIHByb3BlcnR5fSBmcm9tICdsaXQvZGVjb3JhdG9ycy5qcyc7XG5pbXBvcnQge2xpdmV9IGZyb20gJ2xpdC9kaXJlY3RpdmVzL2xpdmUuanMnO1xuXG5AY3VzdG9tRWxlbWVudCgnbnVtYmVyLWVsZW1lbnQnKVxuZXhwb3J0IGNsYXNzIE51bWJlckVsZW1lbnRDb21wb25lbnQgZXh0ZW5kcyBMaXRFbGVtZW50IHtcbiAgICBcbiAgQHByb3BlcnR5KHt0eXBlOiBOdW1iZXJ9KVxuICB2YWx1ZTogbnVtYmVyIHwgbnVsbDtcblxuICBwdWJsaWMgcmVuZGVyKCkge1xuICAgIHJldHVybiBodG1sYFxuICAgICAgPGlucHV0IFxuICAgICAgICB0eXBlPVwibnVtYmVyXCIgXG4gICAgICAgLnZhbHVlQXNOdW1iZXI9JHtsaXZlKHRoaXMudmFsdWUgPz8gbm9DaGFuZ2UpfVxuICAgICAgIEBpbnB1dD0ke3RoaXMuI2hhbmRsZUlucHV0fT5cbiAgICBgO1xuICB9XG5cbiAgI2hhbmRsZUlucHV0KGU6IEV2ZW50KSB7XG4gICAgY29uc3QgdmFsdWUgPSAoZS5jdXJyZW50VGFyZ2V0IGFzIEhUTUxJbnB1dEVsZW1lbnQpLnZhbHVlQXNOdW1iZXI7XG4gICAgdGhpcy52YWx1ZSA9IGlzTmFOKHZhbHVlKSA_IG51bGwgOiB2YWx1ZTtcbiAgfVxufVxuXG5kZWNsYXJlIGdsb2JhbCB7XG4gIGludGVyZmFjZSBIVE1MRWxlbWVudFRhZ05hbWVNYXAge1xuICAgICdudW1iZXItZWxlbWVudCc6IE51bWJlckVsZW1lbnRDb21wb25lbnQ7XG4gIH1cbn1cbiJ9LHsibmFtZSI6ImluZGV4Lmh0bWwiLCJjb250ZW50IjoiPCFET0NUWVBFIGh0bWw-XG48aGVhZD5cbiAgPHNjcmlwdCB0eXBlPVwibW9kdWxlXCIgc3JjPVwiLi9yZXByby5qc1wiPjwvc2NyaXB0PlxuPC9oZWFkPlxuPGJvZHk-XG4gIDxudW1iZXItZWxlbWVudCB2YWx1ZT0xMjM-PC9udW1iZXItZWxlbWVudD5cbjwvYm9keT5cbiJ9LHsibmFtZSI6InBhY2thZ2UuanNvbiIsImNvbnRlbnQiOiJ7XG4gIFwiZGVwZW5kZW5jaWVzXCI6IHtcbiAgICBcImxpdFwiOiBcIl4yLjAuMFwiLFxuICAgIFwiQGxpdC9yZWFjdGl2ZS1lbGVtZW50XCI6IFwiXjEuMC4wXCIsXG4gICAgXCJsaXQtZWxlbWVudFwiOiBcIl4zLjAuMFwiLFxuICAgIFwibGl0LWh0bWxcIjogXCJeMi4wLjBcIlxuICB9XG59IiwiaGlkZGVuIjp0cnVlfV0).

If you make your component fully controlled, then you would alway set the value, and either keep the current string value around to set into the input, _or_ add a keydown event handler to prevent entering non-valid characters in the first place. The fact that `<input>` allows dashes and dots in invalid locations is somewhat surprising. It's a bit more complicated to do this because you want to allow dashes at the start of the value, and `input.selectionStart` doesn't work for number inputs (why??), so you have to use a plain input. You can see that [here](https://lit.dev/playground/#project=W3sibmFtZSI6InJlcHJvLnRzIiwiY29udGVudCI6ImltcG9ydCB7aHRtbCwgTGl0RWxlbWVudCwgbm9DaGFuZ2V9IGZyb20gJ2xpdCc7XG5pbXBvcnQge2N1c3RvbUVsZW1lbnQsIHByb3BlcnR5LCBzdGF0ZX0gZnJvbSAnbGl0L2RlY29yYXRvcnMuanMnO1xuaW1wb3J0IHtsaXZlfSBmcm9tICdsaXQvZGlyZWN0aXZlcy9saXZlLmpzJztcblxuQGN1c3RvbUVsZW1lbnQoJ251bWJlci1lbGVtZW50JylcbmV4cG9ydCBjbGFzcyBOdW1iZXJFbGVtZW50Q29tcG9uZW50IGV4dGVuZHMgTGl0RWxlbWVudCB7XG4gICAgXG4gIEBwcm9wZXJ0eSh7dHlwZTogTnVtYmVyfSlcbiAgdmFsdWU6IG51bWJlciB8IG51bGw7XG4gIFxuICBwdWJsaWMgcmVuZGVyKCkge1xuICAgIHJldHVybiBodG1sYFxuICAgICAgPGlucHV0IFxuICAgICAgIC52YWx1ZT0ke3RoaXMudmFsdWV9XG4gICAgICAgQGlucHV0PSR7dGhpcy4jb25JbnB1dH1cbiAgICAgICBAa2V5ZG93bj0ke3RoaXMuI29uS2V5ZG93bn0-XG4gICAgYDtcbiAgfVxuICBcbiAgI29uS2V5ZG93bihlOiBLZXlib2FyZEV2ZW50KSB7XG4gICAgY29uc3Qge3NlbGVjdGlvblN0YXJ0LCB2YWx1ZX0gPSAoZS50YXJnZXQgYXMgSFRNTElucHV0RWxlbWVudCk7XG4gICAgY29uc3Qge2tleX0gPSBlO1xuICAgIGNvbnN0IHZhbGlkS2V5ID1cbiAgICAgICAgL1swLTldLy50ZXN0KGUua2V5KSB8fFxuICAgICAgICBlLmtleS5sZW5ndGggPiAxIHx8XG4gICAgICAgIGUua2V5ID09PSAnLScgJiYgc2VsZWN0aW9uU3RhcnQgPT09IDAgfHxcbiAgICAgICAgZS5rZXkgPT09ICcuJyAmJiAhdmFsdWUuaW5jbHVkZXMoJy4nKVxuICAgIGlmICghdmFsaWRLZXkpIHtcbiAgICAgIGUucHJldmVudERlZmF1bHQoKTtcbiAgICB9XG4gIH1cblxuICAjb25JbnB1dChlOiBFdmVudCkge1xuICAgIGNvbnN0IHZhbHVlU3RyaW5nID0gKGUuY3VycmVudFRhcmdldCBhcyBIVE1MSW5wdXRFbGVtZW50KS52YWx1ZTtcbiAgICBjb25zdCB2YWx1ZSA9IE51bWJlcih2YWx1ZVN0cmluZyk7XG4gICAgdGhpcy52YWx1ZSA9IGlzTmFOKHZhbHVlKSA_IG51bGwgOiB2YWx1ZTtcbiAgfVxufVxuXG5kZWNsYXJlIGdsb2JhbCB7XG4gIGludGVyZmFjZSBIVE1MRWxlbWVudFRhZ05hbWVNYXAge1xuICAgICdudW1iZXItZWxlbWVudCc6IE51bWJlckVsZW1lbnRDb21wb25lbnQ7XG4gIH1cbn1cbiJ9LHsibmFtZSI6ImluZGV4Lmh0bWwiLCJjb250ZW50IjoiPCFET0NUWVBFIGh0bWw-XG48aGVhZD5cbiAgPHNjcmlwdCB0eXBlPVwibW9kdWxlXCIgc3JjPVwiLi9yZXByby5qc1wiPjwvc2NyaXB0PlxuPC9oZWFkPlxuPGJvZHk-XG4gIDxudW1iZXItZWxlbWVudCB2YWx1ZT0xMjM-PC9udW1iZXItZWxlbWVudD5cbjwvYm9keT5cbiJ9LHsibmFtZSI6InBhY2thZ2UuanNvbiIsImNvbnRlbnQiOiJ7XG4gIFwiZGVwZW5kZW5jaWVzXCI6IHtcbiAgICBcImxpdFwiOiBcIl4yLjAuMFwiLFxuICAgIFwiQGxpdC9yZWFjdGl2ZS1lbGVtZW50XCI6IFwiXjEuMC4wXCIsXG4gICAgXCJsaXQtZWxlbWVudFwiOiBcIl4zLjAuMFwiLFxuICAgIFwibGl0LWh0bWxcIjogXCJeMi4wLjBcIlxuICB9XG59IiwiaGlkZGVuIjp0cnVlfV0)

**Reply by augustjk** [reply]
I think `<input>` type number allows dashes and dots for negatives and floats. It somehow allows at most two dashes to be present in the field. 🤷

**Reply by Totati** [reply]
Number inputs are quite smart but hardheaded, just don't want to change the selection, decimal separator, read invalid values or set invalid values 😅. I'm against preventing the user from typing something invalid more than the browser, because that invalid value may be 1 character away from a valid value. It's not possible to differentiate a miss-formatted value from a missing value based on the value of the `value` or `valueAsNumber` property, because the first one returns `''` the second one `NaN` for both. However these can be worked out using the validity property of the input.
* User did not fill the input? You can find out checking the `input.validity.valueMissing` property.
* User typed invalid values? You can find out checking the `input.validity.badInput`. (It's not accident number inputs allow two `-` and an `e` character, `-2e-2` is a valid number and `Number('-2e-2')` will parse it as well)
* You want to allow decimals only? Don't set step, or set it to `1` and check the `input.validity.stepMismatch`
* You want to allow any number with any precision? Set `[step="any"]`
* You want to allow 2 decimal places? Set `[step="0.01"] and check the `input.validity.stepMismatch`

[My playground](https://stackblitz.com/edit/js-s9lixv?file=index.js)

**Reply by justinfagnani** [reply]
Yeah, if you don't even want to restrict what the user can type, then even intput/type=number might not be the best choice. Just use a plain input and figure out your strategy representing a value in your custom element for when the string is invalid.

**Comment by Benexx** [comment]
Using `noChange` allows us to manipulate the value when we want without strange behavior for the user. Thanks @Totati for the answer!
